map/client/client.go

```diff
@@ -0,0 +1,335 @@
+	return err
+}
+
+// PayloadSize returns the size of the user payload
+func (c *Client) PayloadSize() int {
+	geo := c.Session.SphinxGeometry()
+	return geo.UserForwardPayloadLength - cborFrameOverhead - eddsa.SignatureSize
+}
+
+// Get requests ID from the chosen storage node and returns a payload or error
+func (c *Client) Get(ID common.MessageID, signature []byte) ([]byte, error) {
+
+	if !ID.ReadPk().Verify(signature, ID[:]) {
+		return nil, errors.New("signature does not verify Read")
+	}
+	loc, err := c.GetStorageProvider(ID)
+	if err != nil {
+		return nil, err
+	}
+	b := &common.MapRequest{ID: ID, Signature: signature}
+	serialized, err := cbor.Marshal(b)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := c.Session.BlockingSendUnreliableMessage(loc.Name(), loc.Provider(), serialized)
+	if err != nil {
+		return nil, err
+	}
+	// unwrap the response and return the payload
+	resp := &common.MapResponse{}
+	err = cbor.Unmarshal(r, resp)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Status == common.StatusNotFound {
+		return nil, ErrStatusNotFound
+	}
+	return resp.Payload, nil
+}
+
+// RWClient has both Get and Put
+type RWClient interface {
+	Put(addr []byte, payload []byte) error
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// ROClient only has Get
+type ROClient interface {
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// WOClient only has Put
+type WOClient interface {
+	Put(addr []byte, payload []byte) error
+	PayloadSize() int
+}
+
+// rwMap implements ReadWriteTranport using a ReadWriteCap and reference to map client
+// The idea is that several streams can share map/client and client/session but
+// have different capabilities for the underlying Streams transported by map/client
+type rwMap struct {
+	c     *Client
+	rwCap common.ReadWriteCap
+}
+
+// Get implements RWClient.Get
+func (r *rwMap) Get(addr []byte) ([]byte, error) {
+	i := r.rwCap.Addr(addr)
+	k := r.rwCap.Read(addr)
+	return r.c.Get(i, k.Sign(i.Bytes()))
+}
+
+// Put implements RWClient.Put
+func (r *rwMap) Put(addr []byte, payload []byte) error {
+	i := r.rwCap.Addr(addr)
+	k := r.rwCap.Read(addr)
+	return r.c.Put(i, k.Sign(payload), payload)
+}
+
+// PayloadSize implements RWClient.PayloadSize
+func (r *rwMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// roMap implements ReadOnlyTranport using a ReadOnlyCap and reference to map client
+type roMap struct {
+	c     *Client
+	roCap common.ReadOnlyCap
+}
+
+// Get implements ROClient.Get
+func (r *roMap) Get(addr []byte) ([]byte, error) {
+	i := r.roCap.Addr(addr)
+	k := r.roCap.Read(addr)
+	return r.c.Get(i, k.Sign(i.Bytes()))
+}
+
+// PayloadSize implements ROClient.PayloadSize
+func (r *roMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// woMap implements WOClient
+type woMap struct {
+	c     *Client
+	woCap common.WriteOnlyCap
+}
+
+// Put implements WOClient.Put
+func (w *woMap) Put(addr []byte, payload []byte) error {
```

Comment:
here's another example of a method that needs to be rewritten so that we can test the math separately from the io